### PR TITLE
Initial pass at a client-side overview doc

### PR DESCRIPTION
**Pull request by davidrapson**
Prompted by @feedmypixel I've written up a high-level document for how the Membership CSS is structured. It's not intended to be thorough documentation, but rather an overview of what conventions we follow. Largely inspired by these [CSS at…](https://css-tricks.com/css/) articles.

Feedback definitely welcome, especially anything that reads as too prescriptive or teaching you to suck eggs. Aiming for something that just outlines the structure of the CSS.

There's a bit of work to follow to slightly restructure the CSS to split out the `helper` classes from `components` to help make the distinction outlined in the doc clearer in the code.

This PR also cleans up the README a bit and extracts the Troubleshooting section into a separate doc.

### CODE REVIEW

**Comment by davidrapson** [comment]
@feedmypixel Mind giving this a once over.

File: docs/FRONTEND.md [hunk 1]
```diff
@@ -0,0 +1,203 @@
...
+* [Naming Conventions](#naming)
+* [BEM Recommendations](#bem)
+* [JavaScript](#javascript)
+* [Vendor Prefixes](#prefixes)
+* [Further Reading](#reading)
+
+<a name="architecture"></a>
+## Architecture
+
+We use Sass for are stylesheets and break styles down into the following groups:
```
Comment (mattandrews) [inline docs/FRONTEND.md]: *our* stylesheets

Comment (davidrapson) [inline docs/FRONTEND.md]: fnark. Good catch.

File: docs/FRONTEND.md [hunk 2]
```diff
@@ -0,0 +1,203 @@
...
+
+## Helpers
+
+Any low-level helper classes or traits use a set of common prefixes. Where possible helper classes should be a single class with no children.
+
+- `.l-` for layout helpers
+- `.h-` for heading traits
+- `.text-` for text helpers
+- `.grid-` for grid helpers
+- `.u-`o-  fr utilities
```
Comment (mattandrews) [inline docs/FRONTEND.md]: *for* utilities (should the `o-` bit be there?)

File: docs/FRONTEND.md [hunk 3]
```diff
@@ -0,0 +1,203 @@
...
+    <img class="tweet__avatar" src="{$src}" alt="{$alt}">
+  </header>
+  <div class="tweet__body">
+  </div>
+</article>
+```
+
+### State Classes
+
+Use `is-state-name` for state-based modifications of components. The state name should be lowercase with multiple words separated with hyphens. **Avoid stling these classes directly; they should be used as an adjoining class.**
```
Comment (mattandrews) [inline docs/FRONTEND.md]: "stling" typo

**Comment by davidrapson** [comment]
Should've proof read this better…

File: docs/FRONTEND.md [hunk 4]
```diff
@@ -0,0 +1,203 @@
...
+      padding: 1%;
+   }
+}
+```
+
+While the former keeps specificity as low as possible, it’s easier to reason about the second level as we are documenting a known set of variants for a component. If you are consistently adding modifiers to child elements it’s a good indicator that the component is too large; that element should probably be the root of a new component.
+
+### Don’t use BEM children out of context.
+
+One risk with BEM naming conventions is that it’s possible to use a child-selector outside of the context of it’s original parent. So if you see something like the following then it’s a sign that your component wasn’t granular enough or you need to refactor your views:
```
Comment (mattandrews) [inline docs/FRONTEND.md]: "it’s original parent" should be "its original parent".

File: docs/FRONTEND.md [hunk 5]
```diff
@@ -0,0 +1,203 @@
...
+## JavaScript
+
+> Separate style and behavior concerns by using `.js-` prefixed classes for behavior.
+
+**`.js-` classes should never appear in your stylesheets.** They are for JavaScript only. Inversely, there is rarely a reason to see presentation classes like `.header-nav-button` in JavaScript. You should use state classes like `.is-state` in your JavaScript and your stylesheets as `.component.is-state`.
+
+<a name="prefixes"></a>
+## Vendor Prefixes
+
+We use [Autoprefixer](https://github.com/postcss/autoprefixer) for vendor prefixes so write any prefixed CSS using the spec version only and the build process will add any requried prefixes. It's suprising how few vendor prefixes a anymore.
```
Comment (mattandrews) [inline docs/FRONTEND.md]: "It's suprising how few vendor prefixes a anymore"

I think you accidentally a word.

**Comment by mattandrews** [comment]
Sorry for typo fascism, this is a great piece of documentation :+1:

**Comment by davidrapson** [comment]
Greatly appreciated, there were some silly typos in here.

File: README.md [hunk 6]
```diff
@@ -86,33 +105,36 @@ idrun
...
+grunt watch
+```
 
-A library of common patterns used accross the membership site is available at [membership.theguardian.com/patterns](https://membership.theguardian.com/patterns)
+## Compile front-end files
 
-When building new components break them down into fragments and include them in the pattern library.
+```
+grunt compile
+```
```
Comment (feedmypixel) [inline README.md]: worth putting the --dev flag in here? other tasks can be determined by examining the grunt file

File: docs/FRONTEND.md [hunk 7]
```diff
@@ -0,0 +1,203 @@
...
+
+<a name="naming"></a>
+## Naming conventions:
+
+We largely use the following naming conventions in our CSS:
+
+- We (sparingly) use low-level helper classes for commmon traits, e.g., headings, utilities
+- We use a version of BEM to distinguish between components
+- We use state classes (e.g., `is-toggled`) to style components affecteder by JavaScript
+
```
Comment (feedmypixel) [inline docs/FRONTEND.md]: I remember this from our conversation, maybe these are just standard CS rules:
- If it is a component in markup then the Bem classes should display this to the reader
- Componentize markup where possible for reuse, until it feels "right" not to big not to small

File: docs/FRONTEND.md [hunk 8]
```diff
@@ -0,0 +1,203 @@
...
+- `.l-` for layout helpers
+- `.h-` for heading traits
+- `.text-` for text helpers
+- `.grid-` for grid helpers
+- `.u-` for utilities
+
+### Layout helpers
+
+Layout helpers are common traits used to extract out common layout traits, for example constraining widths (`l-constrained`) or insetting content (`l-pad`).
+
```
Comment (feedmypixel) [inline docs/FRONTEND.md]: is this now .l-inset ? (I realise I am being a bit picky)

File: docs/FRONTEND.md [hunk 9]
```diff
@@ -0,0 +1,203 @@
...
+The grid helpers provide a common pattern for setting up 2up, 3up and 4up grids. It's not a full grid system but provides a useful abstraction for common equal-width grids. See the events listing pages as an example.
+
+### Utilities
+
+Utilities classe (prefixed with `u-`) are patterns for common overrides, they are included at the end of the stylesheet and typically include `!important` declarations as they act as style trumps. Typical utilities include clearfixes (`u-cf`) and margin resets (`u-no-margin`).
+
+
+## Components
+
+Syntax: `<componentName>[--modifierName|-descendantName]`
```
Comment (feedmypixel) [inline docs/FRONTEND.md]: should this be __descendantName

File: docs/FRONTEND.md [hunk 10]
```diff
@@ -0,0 +1,203 @@
...
+  <p class="urelated-component__child"></p>
+</article>
+```
+
+<a name="javascript"></a>
+## JavaScript
+
+> Separate style and behavior concerns by using `.js-` prefixed classes for behavior.
+
+**`.js-` classes should never appear in your stylesheets.** They are for JavaScript only. Inversely, there is rarely a reason to see presentation classes like `.header-nav-button` in JavaScript. You should use state classes like `.is-state` in your JavaScript and your stylesheets as `.component.is-state`.
```
Comment (feedmypixel) [inline docs/FRONTEND.md]: is it worth also mentioning qa-

**Comment by feedmypixel** [comment]
This is great, really good to see the conversation written down.
A few comments added :+1:

File: docs/FRONTEND.md (hunk 10 above)
Comment (davidrapson) [inline docs/FRONTEND.md]: @feedmypixel Good call. Done.

File: README.md (hunk 6 above)
Comment (davidrapson) [inline README.md]: @feedmypixel Probably one for another PR.

**Pull request merged**